File: Map/VisualStudio/MapDataProcessing/MapElement.cs
/* --------------------------------------------------------------------
 *
 * Copyright (C) 2018
 *
 * This file is part of the Diaphnea project.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * -------------------------------------------------------------------- */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Schemas;

namespace MapDataProcessing
{
    abstract class MapElement
    {
        private readonly String _id;
        protected readonly MapData _mapData;
        private readonly ElementName _name;
        private readonly ElementName _shortName;
        private readonly double _importance;
        private readonly List<Look> _looks = new List<Look>();
        private readonly Category _category;

        private readonly Dictionary<MapElement, int> _linkedElements1 = new Dictionary<MapElement, int>();
        private readonly Dictionary<MapElement, int> _linkedElements2 = new Dictionary<MapElement, int>();
        internal Dictionary<MapElement, int> LinkedElements1 { get { return _linkedElements1; } }
        internal Dictionary<MapElement, int> LinkedElements2 { get { return _linkedElements2; } }
        private readonly Dictionary<string, List<List<string>>> _nameDictionary = new Dictionary<string, List<List<string>>>();

        internal double Importance { get { return _importance; } }
        internal List<Look> Looks { get { return _looks; } }

        internal Dictionary<string, List<List<string>>> NameDictionary { get { return _nameDictionary; } }

        internal MapElement(String id, MapData mapData, XmlName[] name, XmlName[] shortName, double importance, List<string> lookIds, string categoryId)
        {
            _id = id;
            _mapData = mapData;
            _name = new ElementName(name);
            _shortName = new ElementName(shortName);
            _importance = importance;
            _category = mapData.getCategory(categoryId);

            foreach (string lookId in lookIds) _looks.Add(mapData.getLook(lookId));

            if (_name != null) addName(_name);
            if (_shortName != null) addName(_shortName);
        }

        abstract protected bool allowMultiline();
        abstract internal int addKmlFile(String path);
        abstract internal int formParts1();
        abstract internal int formParts2();

        internal string Id { get { return _id; } }
        protected MapData MapData { get { return _mapData; } }
        internal Category Category { get { return _category; } }

        private void addName(ElementName name)
        {
            foreach (KeyValuePair<string, string> pair in name.Dictionary)
            {
                if (pair.Value.Length > 0) addName(pair.Key, pair.Value);
            }
        }

        private void addName(string languageId, string name)
        {
            if (!allowMultiline())
            {
                List<string> lineList = new List<string>();
                lineList.Add(name);

                if (!_nameDictionary.ContainsKey(languageId)) _nameDictionary.Add(languageId, new List<List<string>>());
                _nameDictionary[languageId].Add(lineList);
            }
            else
            {
                bool firstCharFound = false;
                int startIndex = 0;
                List<string> wordList = new List<string>();
                List<int> wordSizeList = new List<int>();

                int i, n = name.Length;
                for (i = 0; i < n; ++i)
                {
                    if (i == n - 1 || (firstCharFound && (name[i] == '-' || name[i] == ' ') && name[i + 1] != '-' && name[i + 1] != ' '))
                    {
                        string word = name.Substring(startIndex, i + 1 - startIndex);
                        wordList.Add(word);
                        wordSizeList.Add(word.Length);
                        startIndex = i + 1;
                    }
                    else if (name[i] != '-' && name[i] != ' ')
                    {
                        firstCharFound = true;
                    }
                }

                n = wordList.Count() - 1;
                int k;
                for (k = 0; k <= n; ++k)
                {
                    // The goal of this algorithm is:
                    // Find the best way to split a (n+1)-word string into k+1 lines.
                    // (The best split is the one that minimizes the text width.
                    // The text width is the width of the longest line)
                    // There are C(n, k) ways to split a (n+1)-word string into k+1 lines
                    // (where C(n, k) is the number of k-combinations from a set of n elements)
                    List<string> bestLineList = null;
                    int bestWidth = Int32.MaxValue;
                    int combinationCount = Combinations.getCount(n, k);
                    for (i = 0; i < combinationCount; ++i)
                    {
                        List<string> lineList = new List<string>();
                        int width = 0;
                        int index0 = 0;
                        int j;
                        for (j = 0; j <= k; ++j)
                        {
                            int index;
                            if (j < k) index = Combinations.get(n, k, i, j);
                            else index = n;

                            int w = 0;
                            string line = "";
                            int p;
                            for (p = index0; p <= index; ++p)
                            {
                                line += wordList[p];
                                w += wordSizeList[p];
                            }

                            while (line[line.Length - 1] == ' ')
                            {
                                line = line.Remove(line.Length - 1);
                                --w;
                            }

                            lineList.Add(line);
                            index0 = index + 1;
                            if (w > width) width = w;
                        }

                        if (width < bestWidth)
                        {
                            bestWidth = width;
                            bestLineList = lineList;
                        }
                    }

                    if (!_nameDictionary.ContainsKey(languageId)) _nameDictionary.Add(languageId, new List<List<string>>());
                    _nameDictionary[languageId].Add(bestLineList);
                }
            }
        }

        internal string getInfoJson()
        {
            List<string> linkedElements1Array = new List<string>();
            foreach (MapElement linkedElement in _linkedElements1.Keys) linkedElements1Array.Add(String.Format("\"{0}\"", linkedElement.Id));
            string linkedElements1 = String.Join(",", linkedElements1Array);

            List<string> linkedElements2Array = new List<string>();
            foreach (MapElement linkedElement in _linkedElements2.Keys) linkedElements2Array.Add(String.Format("\"{0}\"", linkedElement.Id));
            string linkedElements2 = String.Join(",", linkedElements2Array);

            string tmp = String.Format("{{\"names\":{0},\"linkedElements1\":[{1}],\"linkedElements2\":[{2}],\"category\":{3}}}",
                _name.getJson(), linkedElements1, linkedElements2, _category.Id);

            return tmp;
        }

        abstract internal int generateCode(CodeGenerator codeGenerator);
    }
}

File: Map/NodeJS/map_server/Data/MapElement.cpp
#include "MapElement.h"
#include "ElementName.h"
#include "Combinations.h"
#include "IMap.h"

#undef max
#include <limits>
#include <locale>
#include <codecvt>

namespace map_server
{
    MapElement::~MapElement()
    {
        std::map<std::string, std::vector<ElementName *> >::iterator it = _nameMap.begin();
        for (; it != _nameMap.end(); ++it)
        {
            int i, n = (*it).second.size();
            for (i = 0; i < n; ++i) delete (*it).second[i];
        }
    }

    bool MapElement::loadCommon(mongo::BSONObj dbElement)
    {
        mongo::BSONElement importanceElt = dbElement.getField("importance");
        mongo::BSONElement nameElt = dbElement.getField("name");
        mongo::BSONElement shortNameElt = dbElement.getField("short_name");
        int categoryId = dbElement.getIntField("category_id");

        if (importanceElt.type() != mongo::NumberDouble || nameElt.type() != mongo::Object || shortNameElt.type() != mongo::Object || categoryId < 0 || categoryId > _maxIntDbValue)
        {
            return false;
        }

        _importance = importanceElt.Double();
        _category = _iMap->getCategory(categoryId);

        mongo::BSONObj dbName = nameElt.Obj();
        mongo::BSONObj dbShortName = shortNameElt.Obj();

        std::string namesJson;

        int i, n = _iMap->getLanguageIdVectorPtr()->size();
        for (i = 0; i < n; ++i)
        {
            const char *languageId = (*_iMap->getLanguageIdVectorPtr())[i].c_str();
            std::string name = dbName.getStringField(languageId);
            std::string shortName = dbShortName.getStringField(languageId);
            // Empty 'name' and/or 'shortName' is not an error

            std::vector<ElementName *> nameVector;
            addNames(name, nameVector);
            addNames(shortName, nameVector);

            _nameMap.insert(std::pair<std::string, std::vector<ElementName *> >(languageId, nameVector));

            if (namesJson.empty()) namesJson = "{\"";
            else namesJson += "\",\"";
            namesJson += std::string(languageId) + "\":\"" + name;
        }
        if (!namesJson.empty()) namesJson += "\"}";
        else namesJson = "null";

        std::string linkedElements1Json;
        mongo::BSONElement linkedElements1Elt = dbElement.getField("linked_elements1");
        if (linkedElements1Elt.type() != mongo::Array) return false;
        std::vector<mongo::BSONElement> dbLinkedElement1Vector = linkedElements1Elt.Array();

        n = dbLinkedElement1Vector.size();
        for (i = 0; i < n; ++i)
        {
            mongo::BSONElement linkedElement1Elt = dbLinkedElement1Vector[i];
            if (linkedElement1Elt.type() != mongo::String) return false;
            std::string linkedElement1 = linkedElement1Elt.String();
            if (linkedElement1.empty()) return false;

            if (linkedElements1Json.empty()) linkedElements1Json = "[\"";
            else linkedElements1Json += "\",\"";
            linkedElements1Json += linkedElement1;
        }
        if (!linkedElements1Json.empty()) linkedElements1Json += "\"]";
        else linkedElements1Json = "[]";

        std::string linkedElements2Json;
        mongo::BSONElement linkedElements2Elt = dbElement.getField("linked_elements2");
        if (linkedElements2Elt.type() != mongo::Array) return false;
        std::vector<mongo::BSONElement> dbLinkedElement2Vector = linkedElements2Elt.Array();

        n = dbLinkedElement2Vector.size();
        for (i = 0; i < n; ++i)
        {
            mongo::BSONElement linkedElement2Elt = dbLinkedElement2Vector[i];
            if (linkedElement2Elt.type() != mongo::String) return false;
            std::string linkedElement2 = linkedElement2Elt.String();
            if (linkedElement2.empty()) return false;

            if (linkedElements2Json.empty()) linkedElements2Json = "[\"";
            else linkedElements2Json += "\",\"";
            linkedElements2Json += linkedElement2;
        }
        if (!linkedElements2Json.empty()) linkedElements2Json += "\"]";
        else linkedElements2Json = "[]";

        _infoJson = "{\"names\":" + namesJson + ",\"linkedElements1\":" + linkedElements1Json + ",\"linkedElements2\":" + linkedElements2Json
                    + ",\"category\":" + std::to_string(categoryId) + "}";

        return true;
    }

    void MapElement::addNames(const std::string& name, std::vector<ElementName *>& nameVector)
    {
        if (!name.empty())
        {
            if (!allowMultiline())
            {
                std::vector<std::string> lineVector;
                lineVector.push_back(name);
                nameVector.push_back(new ElementName(lineVector));
            }
            else
            {
                bool firstCharFound = false;
                int startIndex = 0;
                std::vector<std::string> wordVector;
                std::vector<int> wordSizeVector;

                std::wstring_convert<std::codecvt_utf8<wchar_t> > converter;

                int i, n = name.size();
                for (i = 0; i < n; ++i)
                {
                    if (i == n - 1 || (firstCharFound && (name[i] == '-' || name[i] == ' ') && name[i + 1] != '-' && name[i + 1] != ' '))
                    {
                        std::string word = name.substr(startIndex, i + 1 - startIndex);
                        wordVector.push_back(word);
                        wordSizeVector.push_back(converter.from_bytes(word).size());
                        startIndex = i + 1;
                    }
                    else if (name[i] != '-' && name[i] != ' ')
                    {
                        firstCharFound = true;
                    }
                }

                n = wordVector.size() - 1;
                int k;
                for (k = 0; k <= n; ++k)
                {
                    // The goal of this algorithm is:
                    // Find the best way to split a (n+1)-word string into k+1 lines.
                    // (The best split is the one that minimizes the text width.
                    // The text width is the width of the longest line)
                    // There are C(n, k) ways to split a (n+1)-word string into k+1 lines
                    // (where C(n, k) is the number of k-combinations from a set of n elements)
                    std::vector<std::string> bestLineVector;
                    int bestWidth = std::numeric_limits<int>::max();
                    int combinationCount = Combinations::getCount(n, k);
                    for (i = 0; i < combinationCount; ++i)
                    {
                        std::vector<std::string> lineVector;
                        int width = 0;
                        int index0 = 0;
                        int j;
                        for (j = 0; j <= k; ++j)
                        {
                            int index;
                            if (j < k) index = Combinations::get(n, k, i, j);
                            else index = n;

                            int w = 0;
                            std::string line;
                            int p;
                            for (p = index0; p <= index; ++p)
                            {
                                line += wordVector[p];
                                w += wordSizeVector[p];
                            }

                            while (line.back() == ' ')
                            {
                                line.pop_back();
                                --w;
                            }

                            lineVector.push_back(line);
                            index0 = index + 1;
                            if (w > width) width = w;
                        }

                        if (width < bestWidth)
                        {
                            bestWidth = width;
                            bestLineVector = lineVector;
                        }
                    }

                    nameVector.push_back(new ElementName(bestLineVector));
                }
            }
        }
    }
}
